pkg/orm/client.go

```diff
@@ -0,0 +1,629 @@
+	}
+
+	return nil
+}
+
+// DeleteProject delete the model.ProjectInfo
+func (c *metaOpsClient) DeleteProject(ctx context.Context, projectID string) error {
+	if result := c.db.Where("id=?", projectID).Delete(&model.ProjectInfo{}); result.Error != nil {
+		return cerrors.ErrMetaOpFail.Wrap(result.Error)
+	}
+
+	return nil
+}
+
+// QueryProject query all projects
+func (c *metaOpsClient) QueryProjects(ctx context.Context) ([]*model.ProjectInfo, error) {
+	var projects []*model.ProjectInfo
+	if result := c.db.Find(&projects); result.Error != nil {
+		return nil, cerrors.ErrMetaOpFail.Wrap(result.Error)
+	}
+
+	return projects, nil
+}
+
+// GetProjectByID query project by projectID
+func (c *metaOpsClient) GetProjectByID(ctx context.Context, projectID string) (*model.ProjectInfo, error) {
+	var project model.ProjectInfo
+	if result := c.db.Where("id = ?", projectID).First(&project); result.Error != nil {
+		if result.Error == gorm.ErrRecordNotFound {
+			return nil, cerrors.ErrMetaEntryNotFound.Wrap(result.Error)
+		}
+
+		return nil, cerrors.ErrMetaOpFail.Wrap(result.Error)
+	}
+
+	return &project, nil
+}
+
+// CreateProjectOperation insert the operation
+func (c *metaOpsClient) CreateProjectOperation(ctx context.Context, op *model.ProjectOperation) error {
+	if op == nil {
+		return cerrors.ErrMetaOpFail.GenWithStackByArgs("input project operation is nil")
+	}
+
+	if result := c.db.Create(op); result.Error != nil {
+		return cerrors.ErrMetaOpFail.Wrap(result.Error)
+	}
+
+	return nil
+}
+
+// QueryProjectOperations query all operations of the projectID
+func (c *metaOpsClient) QueryProjectOperations(ctx context.Context, projectID string) ([]*model.ProjectOperation, error) {
+	var projectOps []*model.ProjectOperation
+	if result := c.db.Where("project_id = ?", projectID).Find(&projectOps); result.Error != nil {
+		return nil, cerrors.ErrMetaOpFail.Wrap(result.Error)
+	}
+
+	return projectOps, nil
+}
+
+// QueryProjectOperationsByTimeRange query project operation betweem a time range of the projectID
+func (c *metaOpsClient) QueryProjectOperationsByTimeRange(ctx context.Context,
+	projectID string, tr TimeRange,
+) ([]*model.ProjectOperation, error) {
+	var projectOps []*model.ProjectOperation
+	if result := c.db.Where("project_id = ? AND created_at >= ? AND created_at <= ?", projectID, tr.start,
+		tr.end).Find(&projectOps); result.Error != nil {
+		return nil, cerrors.ErrMetaOpFail.Wrap(result.Error)
+	}
+
+	return projectOps, nil
+}
+
+/////////////////////////////// Job Operation
+// UpsertJob upsert the jobInfo
+// TODO: refine me
+func (c *metaOpsClient) UpsertJob(ctx context.Context, job *libModel.MasterMetaKVData) error {
+	if job == nil {
+		return cerrors.ErrMetaOpFail.GenWithStackByArgs("input master meta is nil")
+	}
+
+	if err := c.db.Create(job).Error; err != nil {
+		if !isDuplicateEntryErr(err) {
+			return cerrors.ErrMetaOpFail.Wrap(err)
+		}
+		if err := c.UpdateJob(ctx, job); err != nil {
+			return err
+		}
+	}
```

Comment:
Can we use the [upsert on conflict](https://gorm.io/docs/create.html#Upsert-On-Conflict) feature that is provided in gorm. Ditto for other upsert implement.

Reply:
I try the `upsert`. 
It seems more proper implement for this interface.
thanks.